src/blocks/calendar.rs

```diff
@@ -0,0 +1,385 @@
+//! `events_within_hours` | Number of hours to look for events in the future | `24`
+//! `warning_threshold` | Warning threshold in seconds for the upcoming event | `300`
+//! `browser_cmd` | Command to open event details in a browser. The block passes the HTML link as an argument | `"xdg-open"`
+//!
+//! Action          | Description                               | Default button
+//! ----------------|-------------------------------------------|---------------
+//! `open_link` | Opens the HTML link of the event | Left
+//!
+//!//! # Examples
+//!
+//! ## Unauthenticated
+//!
+//! ```toml
+//! [[block]]
+//! block = "calendar"
+//! next_event_format = " $icon $start.datetime(f:'%a %H:%M') $summary "
+//! ongoing_event_format = " $icon $summary (ends at $end.datetime(f:'%H:%M')) "
+//! no_events_format = " $icon no events "
+//! interval = 30
+//! url = "https://caldav.example.com/calendar/"
+//! calendars = ["user/calendar"]
+//! events_within_hours = 48
+//! ignore_all_day_events = true
+//! warning_threshold = 600
+//! browser_cmd = "firefox"
+//! [block.auth]
+//! type = "unauthenticated"
+//! ```
+//!
+//! ## Basic Authentication
+//!
+//! ```toml
+//! [[block]]
+//! block = "calendar"
+//! next_event_format = " $icon $start.datetime(f:'%a %H:%M') $summary "
+//! ongoing_event_format = " $icon $summary (ends at $end.datetime(f:'%H:%M')) "
+//! no_events_format = " $icon no events "
+//! interval = 30
+//! url = "https://caldav.example.com/calendar/"
+//! calendars = [ "Holidays" ]
+//! events_within_hours = 48
+//! ignore_all_day_events = true
+//! warning_threshold = 600
+//! browser_cmd = "firefox"
+//! [block.auth]
+//! type = "basic"
+//! username = "your_username"
+//! password = "your_password"
+//! ```
+//! Note: The `username` and `password` can also be provided by setting the environment variables `I3RS_CALENDAR_AUTH_USERNAME` and `I3RS_CALENDAR_PASSWORD` respectively.
+//!
+//! ## OAuth2 Authentication (Google Calendar)
+//!
+//! ```toml
+//! [[block]]
+//! block = "calendar"
+//! next_event_format = " $icon $start.datetime(f:'%a %H:%M') $summary "
+//! ongoing_event_format = " $icon $summary (ends at $end.datetime(f:'%H:%M')) "
+//! no_events_format = " $icon no events "
+//! interval = 30
+//! url = "https://www.googleapis.com/calendar/v3/calendars/primary/events"
+//! calendars = ["primary"]
+//! events_within_hours = 48
+//! ignore_all_day_events = true
+//! warning_threshold = 600
+//! browser_cmd = "firefox"
+//! [block.auth]
+//! type = "oauth2"
+//! client_id = "your_client_id"
+//! client_secret = "your_client_secret"
+//! auth_url = "https://accounts.google.com/o/oauth2/auth"
+//! token_url = "https://oauth2.googleapis.com/token"
+//! auth_token = "~/.config/i3status-rust/calendar.auth_token"
+//! redirect_port = 8080
+//! scopes = ["https://www.googleapis.com/auth/calendar", "https://www.googleapis.com/auth/calendar.events"]
```

Comment:
BTW, I would have thought that the readonly versions of the scopes would be valid. However that doesn't seem to be the case. I opened an issue about it: https://issuetracker.google.com/issues/342730758